zerver/views/streams.py

```diff
@@ -1108,11 +1108,33 @@ def get_stream_email_address(
     user_profile: UserProfile,
     *,
     stream_id: Annotated[NonNegativeInt, ApiParamConfig("stream", path_only=True)],
+    sender_id: Json[NonNegativeInt] | None = None,
 ) -> HttpResponse:
     (stream, sub) = access_stream_by_id(
         user_profile,
         stream_id,
     )
-    stream_email = encode_email_address(stream, show_sender=True)
+    email_gateway_bot = get_system_bot(settings.EMAIL_GATEWAY_BOT, stream.realm_id)
+
+    if sender_id is None:
+        sender = email_gateway_bot
+    elif sender_id == user_profile.id:
+        sender = user_profile
+    else:
+        sender = get_user_profile_by_id(sender_id)
+
+    if (
+        sender.id not in [user_profile.id, email_gateway_bot.id]
+        and sender.bot_owner_id != user_profile.id
+    ):
+        raise JsonableError(
+            _(
+                "The sender ID must be either the current user's ID, the email gateway bot's ID, or the ID of a bot owned by the user."
+            )
```

Comment:
Probably we can use a more generic error message, like "Invalid sender ID". The API documentation explains what the valid values are, and it's often a burden to translators to have very long strings like this unnecessarily.

Reply:
Makes sense. Changed it to `"Invalid sender ID"`